**Lab 5 Unit 4**

For this lab, we will be using the same dataset we used in the previous labs. We recommend using the same notebook since you will be reusing the same variables you previous created and used in labs.

**Instructions**
1. In this final lab, we will model our data. Import sklearn train_test_split and separate the data.
2. Try a simple linear regression with all the data to see whether we are getting good results.
3. Great! Now define a function that takes a list of models and train (and tests) them so we can try a lot of them without repeating code.
4. Use the function to check LinearRegressor and KNeighborsRegressor.
5. You can check also the MLPRegressor for this task!
6. Check and discuss the results.

In [3]:
import pandas as pd
import numpy as np
import datetime
import warnings
import matplotlib.pyplot as plt
import seaborn as sns

pd.set_option('display.max_columns', None)
warnings.filterwarnings('ignore')

In [4]:
customer_df = pd.read_csv("/Users/clark/Desktop/ironhack/labs/unit 4/lab-cleaning-numerical-data/files_for_lab/we_fn_use_c_marketing_customer_value_analysis.csv")
customer_df

Customer       State  Customer Lifetime Value Response  Coverage  \
0     BU79786  Washington              2763.519279       No     Basic   
1     QZ44356     Arizona              6979.535903       No  Extended   
2     AI49188      Nevada             12887.431650       No   Premium   
3     WW63253  California              7645.861827       No     Basic   
4     HB64268  Washington              2813.692575       No     Basic   
...       ...         ...                      ...      ...       ...   
9129  LA72316  California             23405.987980       No     Basic   
9130  PK87824  California              3096.511217      Yes  Extended   
9131  TD14365  California              8163.890428       No  Extended   
9132  UP19263  California              7524.442436       No  Extended   
9133  Y167826  California              2611.836866       No  Extended   

     Education Effective To Date EmploymentStatus Gender  Income  \
0     Bachelor           2/24/11         Employed      F   56274   
1     Bachelor           1/31/11       Unemployed      F       0   
2     Bachelor           2/19/11         Employed      F   48767   
3     Bachelor           1/20/11       Unemployed      M       0   
4     Bachelor            2/3/11         Employed      M   43836   
...        ...               ...              ...    ...     ...   
9129  Bachelor           2/10/11         Employed      M   71941   
9130   College           2/12/11         Employed      F   21604   
9131  Bachelor            2/6/11       Unemployed      M       0   
9132   College            2/3/11         Employed      M   21941   
9133   College           2/14/11       Unemployed      M       0   

     Location Code Marital Status  Monthly Premium Auto  \
0         Suburban        Married                    69   
1         Suburban         Single                    94   
2         Suburban        Married                   108   
3         Suburban        Married                   106   
4            Rural         Single                    73   
...            ...            ...                   ...   
9129         Urban        Married                    73   
9130      Suburban       Divorced                    79   
9131      Suburban         Single                    85   
9132      Suburban        Married                    96   
9133      Suburban         Single                    77   

      Months Since Last Claim  Months Since Policy Inception  \
0                          32                              5   
1                          13                             42   
2                          18                             38   
3                          18                             65   
4                          12                             44   
...                       ...                            ...   
9129                       18                             89   
9130                       14                             28   
9131                        9                             37   
9132                       34                              3   
9133                        3                             90   

      Number of Open Complaints  Number of Policies     Policy Type  \
0                             0                   1  Corporate Auto   
1                             0                   8   Personal Auto   
2                             0                   2   Personal Auto   
3                             0                   7  Corporate Auto   
4                             0                   1   Personal Auto   
...                         ...                 ...             ...   
9129                          0                   2   Personal Auto   
9130                          0                   1  Corporate Auto   
9131                          3                   2  Corporate Auto   
9132                          0                   3   Personal Auto   
9133                          0                   1 

In [5]:
customer_df.columns = customer_df.columns.str.lower().str.replace(" ", "_")
customer_df.columns = customer_df.columns.str.replace("employmentstatus","employment_status")
customer_df.columns = customer_df.columns.str.replace("customer","customer_id")
customer_df.head()

customer_id       state  customer_id_lifetime_value response  coverage  \
0     BU79786  Washington                 2763.519279       No     Basic   
1     QZ44356     Arizona                 6979.535903       No  Extended   
2     AI49188      Nevada                12887.431650       No   Premium   
3     WW63253  California                 7645.861827       No     Basic   
4     HB64268  Washington                 2813.692575       No     Basic   

  education effective_to_date employment_status gender  income location_code  \
0  Bachelor           2/24/11          Employed      F   56274      Suburban   
1  Bachelor           1/31/11        Unemployed      F       0      Suburban   
2  Bachelor           2/19/11          Employed      F   48767      Suburban   
3  Bachelor           1/20/11        Unemployed      M       0      Suburban   
4  Bachelor            2/3/11          Employed      M   43836         Rural   

  marital_status  monthly_premium_auto  months_since_last_claim  \
0        Married                    69                       32   
1         Single                    94                       13   
2        Married                   108                       18   
3        Married                   106                       18   
4         Single                    73                       12   

   months_since_policy_inception  number_of_open_complaints  \
0                              5                          0   
1                             42                          0   
2                             38                          0   
3                             65                          0   
4                             44                          0   

   number_of_policies     policy_type        policy renew_offer_type  \
0                   1  Corporate Auto  Corporate L3           Offer1   
1                   8   Personal Auto   Personal L3           Offer3   
2                   2   Personal Auto   Personal L3           Offer1   
3                   7  Corporate Auto  Corporate L2           Offer1   
4                   1   Personal Auto   Personal L1           Offer1   

  sales_channel  total_claim_amount  vehicle_class vehicle_size  
0         Agent          384.811147   Two-Door Car      Medsize  
1         Agent         1131.464935  Four-Door Car      Medsize  
2         Agent          566.472247   Two-Door Car      Medsize  
3   Call Center          529.881344            SUV      Medsize  
4         Agent          138.130879  Four-Door Car      Medsize

In [6]:
customer_df1 = customer_df.copy()

## 1.

In this final lab, we will model our data. Import sklearn train_test_split and separate the data.

In [1]:
from sklearn.model_selection import train_test_split

In [7]:
X = customer_df1.drop(columns=['customer_id', 'total_claim_amount'])
y = customer_df1['total_claim_amount']

In [8]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

## 2.

Try a simple linear regression with all the data to see whether we are getting good results.

In [23]:
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, r2_score, mean_absolute_error
from sklearn.preprocessing import StandardScaler

numerical_df = ['income', 'monthly_premium_auto', 'months_since_last_claim', 'months_since_policy_inception', 'number_of_open_complaints', 'number_of_policies', 'total_claim_amount']

target_variable = 'total_claim_amount'

X = customer_df1[numerical_df]
y = customer_df1[target_variable]

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

linear_reg_model = LinearRegression()
linear_reg_model.fit(X_train_scaled, y_train)

y_pred = linear_reg_model.predict(X_test_scaled)

mse = mean_squared_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)
mae = mean_absolute_error(y_test, y_pred)

print(f'Mean Squared Error: {mse}')
print(f'R-squared Score: {r2}')
print(f'Mean Absolute Error: {mae}')


Mean Squared Error: 7.659666603736771e-25
R-squared Score: 1.0
Mean Absolute Error: 6.187250432244804e-13


## 3.

Great! Now define a function that takes a list of models and train (and tests) them so we can try a lot of them without repeating code.

In [26]:
from sklearn.neighbors import KNeighborsRegressor

def train_test_models(models, X_train, y_train, X_test, y_test):
    results = []

    for model_name, model in models:
        # Train the model
        model.fit(X_train, y_train)

        # Make predictions on the test set
        y_pred = model.predict(X_test)

        # Evaluate the model
        mse = mean_squared_error(y_test, y_pred)
        r2 = r2_score(y_test, y_pred)
        mae = mean_absolute_error(y_test, y_pred)

        results.append({
            'Model': model_name,
            'Mean Squared Error': mse,
            'R-squared Score': r2,
            'Mean Absolute Error': mae
        })

    return results

## 4.

Use the function to check LinearRegressor and KNeighborsRegressor.

In [27]:
models_to_try = [
    ('Linear Regression', LinearRegression()),
    ('KNeighbors Regressor', KNeighborsRegressor())
]

results = train_test_models(models_to_try, X_train_scaled, y_train, X_test_scaled, y_test)

for result in results:
    print(result)

{'Model': 'Linear Regression', 'Mean Squared Error': 7.659666603736771e-25, 'R-squared Score': 1.0, 'Mean Absolute Error': 6.187250432244804e-13}
{'Model': 'KNeighbors Regressor', 'Mean Squared Error': 3462.357896087649, 'R-squared Score': 0.9564543758625106, 'Mean Absolute Error': 40.13981726732349}


## 5.

You can check also the MLPRegressor for this task!

In [28]:
from sklearn.neural_network import MLPRegressor

models_to_try = [
    ('MLP Regressor', MLPRegressor(random_state=42))
]

results = train_test_models(models_to_try, X_train_scaled, y_train, X_test_scaled, y_test)

for result in results:
    print(result)

{'Model': 'MLP Regressor', 'Mean Squared Error': 107.44515429376374, 'R-squared Score': 0.9986486762938177, 'Mean Absolute Error': 7.316000893240854}


## 6.

Check and discuss the results.

**Linear Regression:**

Mean Squared Error (MSE): 7.66e-25
R-squared Score: 1.0
The results for linear regression are almost perfect, with an incredibly low MSE and an R-squared score of 1.0. This suggests that the linear regression model is fitting the data very well. However, such perfect scores might be an indication of overfitting.

**KNeighbors Regressor:**

Mean Squared Error (MSE): 3462.36
R-squared Score: 0.956
The KNeighbors Regressor performs well but not as perfectly as the linear regression model. The MSE is larger, indicating a higher level of prediction error, and the R-squared score is still quite high, suggesting a good fit.

**MLP Regressor (Neural Network):**

Mean Squared Error (MSE): 107.45
R-squared Score: 0.999
The MLP Regressor, a neural network model, also performs very well. The MSE is higher than that of linear regression but still relatively low, and the R-squared score is very close to 1.0, indicating a strong fit.


- The linear regression model appears to have overfit the data, as suggested by the perfect R-squared score and extremely low MSE. It may not generalize well to new, unseen data.

- The KNeighbors Regressor and MLP Regressor seem to provide good performance, with the MLP Regressor having a slight advantage in terms of a lower MSE and a slightly higher R-squared score.